### PR TITLE
Location Events

### PR DESCRIPTION
**Pull request by Half-Shot**
Render and post ``m.location`` events within riot. Uses [Leaflet](https://www.npmjs.com/package/leaflet).

Requires [riot-web#2774](https://github.com/vector-im/riot-web/pull/2774)

I went for OSM for the mapping service, but if anyone has any better ideas it should be easy enough to swap over to that. The other issue is that I'm using the upload icon as a placeholder, and using stock Leaflet icons for the map marker. That will probably need to be changed before this can be merged. 

Other than that, this should be fairly functional.

### CODE REVIEW

File: src/components/structures/UserSettings.js [hunk 1]
```diff
@@ -337,9 +337,6 @@ module.exports = React.createClass({
             },
         */
         ];
-
-        var syncedSettings = UserSettingsStore.getSyncedSettings();
```
Comment (ara4n) [inline src/components/structures/UserSettings.js]: hm, why are you killing the generic user settings UI?

File: src/components/views/dialogs/LocationInputDialog.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    getDefaultProps: function() {
+        return {
+            focus: true,
+            position: null,
+        };
+    },
+
+    componentDidMount: function() {
+      const leafletMap = new Leaflet.Map(this.refs.map);
```
Comment (ara4n) [inline src/components/views/dialogs/LocationInputDialog.js]: please can haz consistent indentation? O:-)

File: src/components/views/messages/MLocationBody.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        return {
+          leafletMap: null,
+          url: null,
+          body: null,
+          mapEnabled: false,
+         };
+    },
+
+    renderMap: function () {
+      const zoom = ZOOM_STREET;
```
Comment (ara4n) [inline src/components/views/messages/MLocationBody.js]: again, 4 spaces for indentation please

File: src/components/views/messages/MLocationBody.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+          // Build URL
+          const parts = content.geo_uri.substr("geo:".length).split(',');
+          const coords = new Leaflet.LatLng(
+            parseFloat(parts[0]),
+            parseFloat(parts[1])
+          );
+          const uri = `https://www.openstreetmap.org/#map=${ZOOM_STREET}/${parts[0]}/${parts[1]}`;
+          this.setState({uri, coords, body: content.body});
+        } else {
+          this.setState({error: "Mising required fields."});
```
Comment (ara4n) [inline src/components/views/messages/MLocationBody.js]: *Missing

File: src/components/views/rooms/MessageComposer.js [hunk 5]
```diff
@@ -87,10 +87,52 @@ export default class MessageComposer extends React.Component {
...
+    onLocationClick(ev) {
+      const matrixClient = MatrixClientPeg.get();
+      const LocationInputDialog = sdk.getComponent("dialogs.LocationInputDialog");
+      const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
+      const LocationEnabled = UserSettingsStore.getSyncedSetting('displayLocationMaps') === true;
+      if (!navigator.geolocation) {
+        Modal.createDialog(
+          ErrorDialog, {
+          title: "Post Location",
+          description: "You either have disabled location in the browser, or it isnt supported.",
```
Comment (ara4n) [inline src/components/views/rooms/MessageComposer.js]: *isn't

File: src/components/views/rooms/MessageComposer.js [hunk 6]
```diff
@@ -274,6 +316,15 @@ export default class MessageComposer extends React.Component {
                 </div>
             );
 
+            const locationButton = (
+                // Add option for either current or specifed location.
+                // TODO: Add icon
```
Comment (ara4n) [inline src/components/views/rooms/MessageComposer.js]: Hm, in practice we probably are going to want a generic + button or something which provides a popup for the more exotic stuff like push-to-talk, locations, etc.  But we can handle that!

File: src/components/structures/UserSettings.js (hunk 1 above)
Comment (ara4n) [inline src/components/structures/UserSettings.js]: i'd expect that supporting location sharing would be precisely the sort of thing that the generic settings code would be good for.  (or possibly hiding it behind a labs option for now)

**Comment by ara4n** [comment]
from quick inspection this is looking insanely cool :) there are few minor cosmeticish things in review comments, one wtf concerning generic settings, and i haven't actually tried using it yet.

One important thing: what is your proposed format for `m.location` events? I'm assuming:

```
content.geo_uri = "geo:...";
content.body = "some kind of label";
...
```

...but do we support formatted bodies here?

The Right solution to this would be to use the long-awaited "human representation of events" stuff to associate formatted bodies with the metadata, i suspect, but happy to ignore this for now; just want it clarified so we can spec it!

**Comment by Half-Shot** [comment]
Okay, that should be all the minor issues taken care of.

The spec is kinda lacking with what I had in mind.

```
"content": {
        "body": "Big Ben, London, UK",
        "geo_uri": "geo:51.5008,0.1247",
        "msgtype": "m.location",
        "thumbnail_info": {
            ....
        },
        "thumbnail_url": "mxc://localhost/FHyPlCeYUSFFxlgbQYZmoEoe"
    },
```

Has a few problems where I'd assume you'd want body and formatted_body to give some sort of representation to clients that do not support ``m.location`` events. Something along the lines of:
```
"body":"Big Ben @ geo:51.5008,0.1247",
"formatted_body":"Big Ben @ <a href='osm.org/blah'>geo:51.5008,0.1247</a>",
```

I think having a separate ``label`` field for clients that *do* support it would be more beneficial rather than trying to parse body. As far as thumbnail goes, I'm not sure how I'm going to handle that yet. The argument kind of goes that if you are aware of m.location events, then you can probably render them clientside with more up to date representation that something stored in a image. 

Up to you if you want to merge this first or change the spec first :)

**Comment by ara4n** [comment]
happy to add it to spec later. the format looks good, although we may wish to finally implement the "human representation of events" stuff from https://docs.google.com/document/d/1cLSlMZ0tulRJM4ki-ImQ_z6DSySQgTRafO2m13RtgZo/edit#heading=h.6azo8zlrfebi to handle thumbnails & body representation etc of the `m.location` events.  Let's hold off on speccing `m.location` until then.

Going to test this locally and then merge so folks can experiment in labs!

**Comment by ara4n** [comment]
Looks like there are some uncommitted resources here... e.g. img/marker-icon.png (or from the other PR, perhaps).

Otherwise, the only other weirdness is the width of the caption entry form on the map preview (but that's just cosmetic).

Are you sure that the thumbnails in the timeline should be interactively zoomable?  They trap the mouse as you scrollwheel the page and start zooming the image rather than the page.  I wonder if opening OSM (or opening a lightbox) would be better when you click on them, to let the user properly navigate...

**Comment by Half-Shot** [comment]
>Looks like there are some uncommitted resources here... e.g. img/marker-icon.png (or from the other PR, perhaps).

Should be in riot-web, but apparently I didn't add them. Will do so

>Otherwise, the only other weirdness is the width of the caption entry form on the map preview (but that's just cosmetic).

Yeah, that and a margin would be nice. I'll see if I can adjust it a bit.

>Are you sure that the thumbnails in the timeline should be interactively zoomable? They trap the mouse as you scrollwheel the page and start zooming the image rather than the page. I wonder if opening OSM (or opening a lightbox) would be better when you click on them, to let the user properly navigate...

What I might end up doing here is thumbnailing the map at the sender end, and then opening a lightbox for an interactive version.

**Comment by ara4n** [comment]
right. for compatibility with other matrix clients I'd definitely suggest including a link to OSM in the caption (both HTML & plaintext for compatibility), and ideally thumbnailing it at the sender as you suggest.

(this is a really good example of why the 'human representation' event stuff is needed, imo, although given the final soln hasn't been picked there yet we can fudge it with body/formatted_body/thumbnail etc for now)

**Comment by Half-Shot** [comment]
Thumbnailing is proving more challenging than anticipated. I originally assumed you could draw any element onto a canvas, but I'm limited to [images and videos](https://developer.mozilla.org/en-US/docs/Web/API/CanvasImageSource) :man_facepalming: .

**Comment by ara4n** [comment]
does OSM not provide a "just give me an image" API?

**Comment by Half-Shot** [comment]
>does OSM not provide a "just give me an image" API?

Only provides tiles :/. It's up to the client to assemble them which is a bit of a pain and why we're using leaflet in the first place :(. Would it be cool if we could conveniently leave off thumbnails for the time being (since they aren't strictly required) until I can figure out what the best solution it?

**Comment by Half-Shot** [comment]
(New commits because my branch got manged after two months :cry: )

Added support for thumbnails for users that don't wish to display maps. We can render them in the browser :)

**Comment by Half-Shot** [comment]
It was brought up that this might cause a lot of problems for OSM if lots of people view a map. While we can somewhat limit this with thumbnails until hover/click, it *might* be a better idea for riot to host it's own tile server?

**Comment by Ralith** [comment]
A few random thoughts:

When sharing a point, the level of precision the user intends may vary wildly. Are they referring to a specific part of south-east Kansas, or the entire united states? Which should be displayed? It might be good to encode a hint regarding this in the message--perhaps a radius around the point that should be enclosed by the view area of a map tool?

Alternatively (or additionally), the event could contain arbitrary geometry, for example using [GeoJSON](http://geojson.org/) (which can be easily fed to most standard GIS or web mapping tools). That way people can send sets of points, polygons, line strings, or whatever else strikes their fancy.

**Comment by Half-Shot** [comment]
>Alternatively, the event could contain arbitrary geometry, for example using GeoJSON (which can be easily fed to most standard GIS or web mapping tools). That way people can send sets of points (favorite restaurants!), polygons (geocache zones!), line strings (travel routes!), or whatever else strikes their fancy.

I like this one more. Seems like a good idea to be included into the spec as an alternative key.

**Comment by Half-Shot** [comment]
https://github.com/matrix-org/matrix-doc/pull/919 will hopefully widen our scope with regards to what we can do with location events :)

**Comment by t3chguy** [comment]
+ Post Location button has same icon as upload file
+ Post location button is shown whether or not location picker is enabled in settings, any reason?
+ bunch of 404s
+ Post Location button in Modal has 404 `/img/marker-icon-2x.png?cache=1495217303216:1 GET https://riot.ovh/img/marker-icon-2x.png?cache=1495217303216 404 ()`

most of these are probably due to it needing merging to current develop

**Comment by t3chguy** [comment]
expects the marker-icon and marker-shadow to be in /img/$1 even if riot is in a subdir

File: src/components/views/rooms/MessageComposer.js [hunk 7]
```diff
@@ -347,15 +339,19 @@ export default class MessageComposer extends React.Component {
...
-                        onClick={this.onLocationClick} title="Post current location">
+                        onClick={this.onLocationClick} title="Post current location test">
                     <TintableSvg src="img/icons-upload.svg" width="35" height="35"/>
                 </div>
             );
 
+            if (!UserSettingsStore.isFeatureEnabled('inline_maps')) {
+                locationButton = null;
+            }
+
```
Comment (t3chguy) [inline src/components/views/rooms/MessageComposer.js]: typically the pattern for these is
```
let locationButton;
if (UserSettingsStore.isFeatureEnabled('inline_maps')) {
    locationButton = ();
}
```
just for the sake of consistency pls

**Comment by pvagner** [comment]
I am not sure this is a proper place for this comment however I would also like to have current location in profile or something very similar. So for example I can request the precise location of my direct chat partner if they allow me to do that. However I don't want all these location requests in the room timeline if possible. I am thinking that it might somehow complement users presence i.e. last seen %(relativeTime)s ago, near %(approxPlace)s or at %(exactSpot)s if allowed.
Some use cases from the wild:
* Show a map of trusted group members if they wish to publish their location
* Being able to check your relatives or friends

Of course this should all be optional with proper privacy controls and warnings. And of course it would be even more usefull if mobile apps were also sending location aware presence events if allowed.

**Comment by almereyda** [comment]
@pvagner indeed this may very well be the wrong place to discuss this. Yet since what you say is most interesting, I guess together with #3545 a geoindex on top of matrix messages, even allowing for NEAR searches would be something additional.

http://owntracks.org/ comes to mind here. We may in the end need a separate issue for a geo-aware indexing of Matrix messages et al. and would like to wait for suggestions of the maintainers to where and how this could be scoped.

**Comment by jfrederickson** [comment]
Got another thing that might be worth keeping in mind for future flexibility: location streaming. Not sure if that deserves another event type or if could be done with m.location events.

**Comment by jryans** [comment]
Thanks for making this contribution a while back. Since the code base has changed since this was opened, it no longer applies cleanly, and I don't think there's a need to keep it open in this state, as we can always find the code here again if needed. If you are still interested in pursuing this feature, please discuss with us in #riot-dev:matrix.org to find a good approach forward.

**Pull request closed without merging**